### DISCUSSION TITLE
Unable to add new form field

### DISCUSSION THREAD

**Discussion by stuartmcgill** [area17/twill#2345] [Q&A] [answered]
I've just started using Twill this week for a new project and am trying to add a new content type. I've been following the [simple page builder guide](https://twillcms.com/guides/page-builder-with-blade/index.html).

After completing that guide, I'm now trying to add my own custom fields to a new content type ('newsItems'). As far as I can tell from [the docs](https://twillcms.com/docs/modules/form-builder.html), all I need to do to add e.g. 'Author' and 'Article date' fields is to update the `getForm()` method:

```php
public function getForm(TwillModelContract $model): Form
    {
        $form = parent::getForm($model);

        $form->add(
            Input::make()->name('description')->label('Description')->translatable()
        );

        $form->add(
            Medias::make()->name('cover')->label('Cover image')
        );

        $form->add(
            DatePicker::make()->name('article_date')->withoutTime()
        );

        $form->add(
            Input::make()->name('author')->label('Author')->translatable()
        );

        $form->add(
            BlockEditor::make()
        );

        return $form;
    }
```

I don't need to create any migrations, the actual data will just be stored in `news_item_revisions.payload` as the latest version?

I'm really struggling to get this to work. The data _occasionally_ gets saved i.e. this query brings back the data I saved:

```sql
SELECT id, JSON_EXTRACT(payload, '$.author'), JSON_EXTRACT(payload, '$.article_date') from news_item_revisions
```

But my custom data (author, article date) is never displayed when I refresh the CMS editor, and so with the next save it gets wiped (unlike the description field, which behaves properly).

![Screenshot from 2023-10-04 17-54-50](https://github.com/area17/twill/assets/30128592/ddebe01b-db3f-4e77-a1e5-0a1f38bdf2ce)

I think I must have missed something pretty fundamental...

**Comment by ifox** [comment, marked as answer]
Hi @stuartmcgill, each field you add at the form level needs a column in your main model table and it needs to be added to the model `$fillable` array as well. When working with blocks, you don't need any migration, but outside of blocks, data is stored using standard Laravel Eloquent models and attributes.

**Reply by stuartmcgill** [reply]
Thank you very much, it's working after adding the column and updating `$fillable`. I was getting confused by the Cover image in the guide which (because it's a related image) doesn't need either of these things.

**Comment by ifox** [comment]
Each field documentation page is indicating the migration needs if any, eg. https://twillcms.com/docs/form-fields/input.html#:~:text=A%20migration%20to,1

**Comment by ifox** [comment]
Fields like medias, files, browsers, and block editors do not need migrations because they use polymorphic tables from Twill (`twill_mediables`, `twill_fileables`, `twill_related`, `twill_blocks`)

**Comment by YahyaS3d** [comment]
Ensure Field Names Match Database:
Verify that the database table associated with your content type (news_item_revisions) has columns named author and article_date to store the custom data.
